[IMP] runbot_gitlab_minimal: Fixes calculated ssh key (#97)

JesusZapata opened the pull request:
Fixes https://github.com/Vauxoo/runbot-addons/issues/97

Commit:
[IMP] runbot_gitlab_minimal: Fixes calculated ssh key (#97)

[FIX] runbot_gitlab_minimal: Fixes pylint check

JesusZapata commented:
@moylop260 
I have tried to put the runbot_gitlab_minimal module to depend on runbot_travis2docker but the method get_ssh_keys of this module is not called

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -113,6 +116,13 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
                         json['base'] = {'ref': json['source_branch']}
+                    if '/commits/' in url:
+                        url = _get_url('/users?search=%s' %
+                                       json['committer_email'], repo.base)
```
What about if 2 users has same email configured?

JesusZapata replied:
I have 2 email registered
![image](https://cloud.githubusercontent.com/assets/1387970/24164567/8b309c6c-0e44-11e7-89d5-63525542aa4f.png)
And all good

moylop260 replied:
You are using 2 emails for 1 user
I'm talking about 2 users with 1 same email configured

I means,
User: JesusZapata
Email: jesus@vauxoo.com

User: JesusZapata2
Email: jesus@vauxoo.com

JesusZapata replied:
I try to register another user with the same email and gitlab show:

![image](https://cloud.githubusercontent.com/assets/1387970/24165621/4198c748-0e47-11e7-9b4e-665b14adb1fe.png)

moylop260 replied:
We tested and this scenario is not possible
Then is good to continue

Commit:
[REF] runbot_travis2docker: The method get_ssh_keys call api to github for calculated a ssh-key

[FIX] runbot_travis2docker: To use getatte instead of hasattr

[FIX] runbot_gitlab_minimal: Remove pylint disable

[FIX] : Miscellaneous corrections

[FIX] runbot_gitlab_minimal: More conditions for check the user of gitlab

[REF] Change the logic of download the ssh-rsa of github and gitlab

[REF] runbot_gitlab_minimal: Move conditions to better place and document the code

[REF] Return early and concatenate \n before join

moylop260 commented on runbot_travis2docker/models/runbot_build.py:
```diff
@@ -367,9 +367,17 @@ def get_ssh_keys(self, cr, uid, build, context=None):
-                url = "https://github.com/%(login)s.keys" % response[own_key]
-                stream = requests.get(url, stream=True)
-                keys += '\n' + stream.text
+                if (hasattr(build.repo_id, 'uses_gitlab') and
```
`if getattr(build.repo_id, 'uses_gitlab', False):`

JesusZapata replied:
@moylop260 
Done

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -192,10 +201,11 @@ def github_status(self, cr, uid, ids, context=None):
 
     @api.multi
-    def _get_branch_url(self):
+    def _get_branch_url(self):  # pylint: disable=missing-return
```
pylint fixes for other PR

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -192,10 +201,11 @@ def github_status(self, cr, uid, ids, context=None):
 
-    branch_url = fields.Char(compute='_get_branch_url')
+    branch_url = fields.Char(compute='_get_branch_url')  \
+        # pylint: disable=method-compute
```
pylint fixes for other PR

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -125,16 +133,17 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
             if not build.repo_id.uses_gitlab:
-                super(RunbotBuild, self).github_status(cr, uid, ids,
-                                                       context=context)
+                return super(RunbotBuild, self).github_status(cr, uid, ids,
```
pylint fixes for other PR

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -125,16 +133,17 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
 
-    def github_status(self, cr, uid, ids, context=None):
+    def github_status(self, cr, uid, ids, context=None):  \
+            # pylint: disable=missing-return
```
pylint fixes for other PR

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -189,7 +197,7 @@ def github_status(self, cr, uid, ids, context=None):
 
 
-class RunbotBranch(models.Model):
+class RunbotBranclh(models.Model):
```
typo

moylop260 commented on runbot_travis2docker/models/runbot_build.py:
```diff
@@ -367,9 +367,16 @@ def get_ssh_keys(self, cr, uid, build, context=None):
+                    ssh_rsa = build.repo_id.github("/users/%s/keys" %
+                                                   response['user_id'])
+                    for rsa in ssh_rsa:
+                        keys += rsa['key']
```
keys += '\n' + rsa['key']

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -113,6 +114,13 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
+                                       json['committer_email'], repo.base)
+                        response = session.get(url)
+                        response.raise_for_status()
+                        json['author'] = response.json()[0]['username']
```
Assign `committer` too

moylop260 replied:
There are cases where the committer is not a user of gitlab
A example is the following commit: https://github.com/Vauxoo/odoo-peru/commit/ae11e24d2d678b327f64637172a88b92eec7526d where the committer `Arturo Flores` don't have github account and we had a TypeError fixed from https://github.com/Vauxoo/runbot-addons/pull/95

Could you review if we will have a special exception for gitlab?

JesusZapata replied:
I made more conditions to avoid an error if the gitlab user does not exist

moylop260 commented on runbot_travis2docker/models/runbot_build.py:
```diff
@@ -367,9 +367,16 @@ def get_ssh_keys(self, cr, uid, build, context=None):
-                keys += '\n' + stream.text
+                if getattr(build.repo_id, 'uses_gitlab', False):
+                    ssh_rsa = build.repo_id.github("/users/%s/keys" %
+                                                   response['user_id'])
```
Where are you processing the committer and the author here?

JesusZapata replied:
@moylop260 
I changed the way to calculate the github keys

moylop260 commented on runbot_travis2docker/models/runbot_build.py:
```diff
@@ -367,9 +367,16 @@ def get_ssh_keys(self, cr, uid, build, context=None):
+                else:
+                    url = ("https://github.com/%(login)s.keys" %
+                           response[own_key])
+                    stream = requests.get(url, stream=True)
```
Use github's api to get keys too here, please

moylop260 commented on runbot_travis2docker/models/runbot_build.py:
```diff
@@ -367,9 +367,16 @@ def get_ssh_keys(self, cr, uid, build, context=None):
-                stream = requests.get(url, stream=True)
-                keys += '\n' + stream.text
+                if getattr(build.repo_id, 'uses_gitlab', False):
+                    ssh_rsa = build.repo_id.github("/users/%s/keys" %
```
Remove the process of gitlab keys here and use a wrapper from github method

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -113,6 +114,17 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
+                        response.raise_for_status()
+                        data = response.json()
+                        if len(data) == 1:
+                            data = data[0]
```
Return a dict-of-dict similar to github

moylop260 commented:
Use 2 commits:
1. Changes for gitlab module
2. Changes for t2d module

And migrate the first one to OCA PR, please

JesusZapata commented:
@moylop260 
Change the logic of download the ssh-rsa of github and gitlab

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -102,17 +109,35 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
                         urls = url.split('/pulls/')
                         url = urls[0] + '/merge_requests?iid=' + urls[1]
                         is_url_merge = True
+                    is_url_keys = url.endswith('.keys')
```
Move `if '/pulls/' in url:` for `get_url` method

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -93,6 +99,7 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
         """
         for repo in self.browse(cr, uid, ids, context=context):
             is_url_merge = False
+            is_url_keys = False
```
Delete useless assigned variables

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -102,17 +109,35 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
+                                data = response.json()
+                                if len(data) == 1:
+                                    data = data[0]
+                                    json[own_key] = {'login': data['username']}
```
`json[own_key] = {'login': len(data) and data[0]['username'] or {}}`

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -17,12 +17,18 @@
         url = url.replace('/repos/', '/projects/')
+        url = url.replace('/commits/', '/repository/commits/')
         url = url.replace(project_name, urllib.quote(project_name, safe=''))
+        if url.endswith('/keys'):
```
Add a comment of url github vs url gitlab

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -17,12 +17,18 @@
 def _get_url(url, base):
     match_object = re.search('([^/]+)/([^/]+)/([^/.]+(.git)?)', base)
     if match_object:
+        prefix = ('https://%s/api/v3%s'
```
Add a comment because if using this replace...

`Because gitlab api v3 requires admin to get keys from api but...`

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -17,12 +17,18 @@
     match_object = re.search('([^/]+)/([^/]+)/([^/.]+(.git)?)', base)
     if match_object:
+        prefix = ('https://%s/api/v3%s'
+                  if not url.endswith('/keys') else 'https://%s%s')
```
github vs gitlab url comment

JesusZapata commented:
@moylop260 
All changes were made
Thanks for your review

moylop260 reviewed (approved):
Could you use 2 commits 1 by module changed?

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -87,32 +100,51 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
                 continue
             try:
                 url = _get_url(url, repo.base)
                 if url:
```
Is possible a return early using `if not url`?

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -87,32 +100,51 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
                         json = json[0]
                         json['head'] = {'ref': json['target_branch']}
                         json['base'] = {'ref': json['source_branch']}
+                    if '/commits/' in url:
```
use `elif`

moylop260 commented on runbot_gitlab_minimal/models/runbot_repo.py:
```diff
@@ -87,32 +100,51 @@ def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
+                                    'login':
+                                    len(data) and data[0]['username'] or {}
+                                }
+                    if is_url_keys:
```
use `elif`

moylop260 commented on runbot_travis2docker/models/runbot_build.py:
```diff
@@ -367,9 +367,9 @@ def get_ssh_keys(self, cr, uid, build, context=None):
-                keys += '\n' + stream.text
+                ssh_rsa = build.repo_id.github('/users/%(login)s/keys' %
+                                               response[own_key])
+                keys += '\n'.join(rsa['key'] for rsa in ssh_rsa)
```
Use `keys += '\n' + '\n'.join...`

JesusZapata commented:
@moylop260 
I made all changes what did you request
